src/Microsoft.AspNet.Antiforgery/DefaultAntiforgeryTokenStore.cs

```diff
@@ -72,23 +72,43 @@ public AntiforgeryToken GetCookieToken(HttpContext httpContext)
                     Resources.FormatAntiforgery_CookieToken_MustBeProvided(_options.CookieName));
             }
 
-            if (!httpContext.Request.HasFormContentType)
+            StringValues requestToken;
+            if (httpContext.Request.HasFormContentType)
             {
                 // Check the content-type before accessing the form collection to make sure
                 // we throw gracefully.
-                throw new InvalidOperationException(
-                    Resources.FormatAntiforgery_FormToken_MustBeProvided(_options.FormFieldName));
+                var form = await httpContext.Request.ReadFormAsync();
+                requestToken = form[_options.FormFieldName];
             }
 
-            var form = await httpContext.Request.ReadFormAsync();
-            var formField = form[_options.FormFieldName];
-            if (string.IsNullOrEmpty(formField))
+            // Fall back to header if the form value is was not provided.
+            if (requestToken.Count == 0 && _options.HeaderName != null)
             {
-                throw new InvalidOperationException(
-                    Resources.FormatAntiforgery_FormToken_MustBeProvided(_options.FormFieldName));
+                requestToken = httpContext.Request.Headers[_options.HeaderName];
+            }
+
+            if (requestToken.Count == 0)
+            {
+                if (_options.HeaderName == null)
+                {
+                    var message = Resources.FormatAntiforgery_FormToken_MustBeProvided(_options.FormFieldName);
+                    throw new InvalidOperationException(message);
+                }
+                else if (!httpContext.Request.HasFormContentType)
+                {
+                    var message = Resources.FormatAntiforgery_HeaderToken_MustBeProvided(_options.HeaderName);
+                    throw new InvalidOperationException(message);
+                }
+                else
+                {
+                    var message = Resources.FormatAntiforgery_RequestToken_MustBeProvided(
+                        _options.FormFieldName,
+                        _options.HeaderName);
+                    throw new InvalidOperationException(message);
+                }
             }
 
-            return new AntiforgeryTokenSet(formField, requestCookie);
+            return new AntiforgeryTokenSet(requestToken, requestCookie);
```

Comment:
This is the new stuff :+1:

Everything else is just renames and updating resources.